runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -304,16 +342,141 @@ protected void addSideInputValue(StreamRecord<RawUnionValue> streamRecord) {
+                ByteBuffer encodedKey = (ByteBuffer) getKeyedStateBackend().getCurrentKey();
+                @SuppressWarnings("ByteBufferBackingArray")
+                ByteArrayInputStream byteStream = new ByteArrayInputStream(encodedKey.array());
+                try {
+                  return keyCoder.decode(byteStream);
+                } catch (IOException e) {
+                  throw new RuntimeException(
+                      String.format(
+                          Locale.ENGLISH, "Failed to decode encoded key: %s", encodedKey));
+                }
+              }
+            });
+    return sdkHarnessRunner;
   }
 
-  private class SdkHarnessDoFnRunner implements DoFnRunner<InputT, OutputT> {
+  @Override
+  public void processWatermark(Watermark mark) throws Exception {
+    // Due to the asynchronous communication with the SDK harness,
+    // a bundle might still be in progress and not all items have
+    // yet been received from the SDk harness. If we just set this
+    // watermark as the new output watermark, we could violate the
+    // order of the records, i.e. pending items in the SDK harness
+    // could become "late" although they were "on time".
+    //
+    // We can solve this problem using one of the following options:
+    //
+    // 1) Finish the current bundle and emit this watermark as the
+    //    new output watermark. Finishing the bundle ensures that
+    //    all the items have been processed by the SDK harness and
+    //    received by the outputQueue (see below), where they will
+    //    have been emitted to the output stream.
+    //
+    // 2) Put a hold on the output watermark for as long as the current
+    //    bundle has not been finished. We have to remember to manually
+    //    finish the bundle in case we receive the final watermark.
+    //    To avoid latency, we should process this watermark again as
+    //    soon as the current bundle is finished.
+    //
+    // Approach 1) is the easiest, yet 2) gives better throughput due
+    // to the bundle getting cut on every watermark. So we have
+    // implemented 2) below.
+    //
+    if (sdkHarnessRunner.isBundleInProgress()) {
+      if (mark.getTimestamp() >= BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()) {
+        invokeFinishBundle();
+      } else {
+        // It is not safe to advance the output watermark yet, so add a hold on the current
+        // output watermark.
+        setPushedBackWatermark(Math.min(currentOutputWatermark, getPushbackWatermarkHold()));
+        sdkHarnessRunner.setBundleFinishedCallback(
+            () -> {
+              try {
+                processWatermark(mark);
+              } catch (Exception e) {
+                throw new RuntimeException(
+                    "Failed to process pushed back watermark after finished bundle.", e);
+              }
+            });
+      }
+    }
+    super.processWatermark(mark);
+  }
+
+  private static class SdkHarnessDoFnRunner<InputT, OutputT>
```

Comment:
What's the reason to change it to static (adds all the constructor clutter)?

Reply:
Yeah, you can probably take the argument either way. I see this in few other places in the Flink runner and I see your point regarding the required dependencies. On the other hand these redundant constructor argument lists and member declarations reduce readability, sometimes make it hard to find what really matters and overall provide little benefit when the class is an internal of the outer class and fully encapsulated. Not a blocker for me though.

Reply:
We could move this out of the operator now which would increase readability. If we don't split it into a separate static class, we wouldn't be able to do that. I like encapsulation and ExecutableStageDoFnOperator doesn't encourage clean design with its heavy use of virtual methods. I think we have to do a separate refactoring to further increase readability.

Reply:
I would agree with that if `SdkHarnessDoFnRunner` was a component with a prospect of being used independently. Instead it is an implementation detail of the operator that is tightly coupled. Ideally the class wouldn't even exist. I would probably go the opposite way and just delegate all methods to the operator, which would remove the clutter can keep the class super small. It could even be an anonymous class.  Again, this is just for sake of discussion and not actionable for this PR.

Reply:
Regarding the operator as a whole and as discussed in the past, constraints are mostly inherited and would need to be addressed by refactoring the base. I'm not sure about the timing for this though, if it should wait until legacy runner goes away (there is still significant work for portable runner to become adequate replacement for Java pipelines).